File: cpp_solutions/DynamicProgramming/Best _Time_to_Buy_and_Sell_Stock II.cpp
/*
Say you have an array for which the ith element is the price of a given stock on day i.

Design an algorithm to find the maximum profit. You may complete as many transactions as you like (ie, buy one and sell one share of the stock multiple times). 
However, you may not engage in multiple transactions at the same time (ie, you must sell the stock before you buy again).
*/

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int total = 0;
        if(!prices.empty()){
        	for (int i = 0; i < prices.size() - 1; ++i)
        	{
        		if(prices[i+1] > prices[i])
        			total +=  prices[i+1]-prices[i];
        	}
        }
        return total;
    }
};

File: python_solutions/DynamicProgramming/Best_Time_to_Buy_and_Sell_Stock_II.py
class Solution:
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        total = 0
        low, high = sys.maxsize, sys.maxsize
        for x in prices:
            if x > high:
                high = x
            else:
                total += high - low
                high, low = x, x
        return total + high - low
        